File: src/ipc/tmdb.ts
import { invoke } from "@tauri-apps/api/core";
import type { TmdbSearchResult, TmdbMovieDetails } from "./types";

export interface TmdbTvSearchResult {
  tmdb_tv_id: number;
  name: string;
  original_name: string;
  first_air_year: number | null;
  overview: string;
  poster_url: string | null;
  number_of_seasons: number | null;
}

export interface TmdbTvSeasonEpisode {
  season_number: number;
  episode_number: number;
  name: string;
  overview: string;
  air_date: string | null;
}

export const tmdbIpc = {
  search: (query: string) =>
    invoke<TmdbSearchResult[]>("tmdb_search", { query }),
  details: (tmdbId: number) =>
    invoke<TmdbMovieDetails>("tmdb_movie_details", { tmdbId }),
  tvSearch: (query: string) =>
    invoke<TmdbTvSearchResult[]>("tmdb_tv_search", { query }),
  tvSeason: (tmdbTvId: number, seasonNumber: number) =>
    invoke<TmdbTvSeasonEpisode[]>("tmdb_tv_season", {
      tmdbTvId,
      seasonNumber,
    }),
};

File: src-tauri/src/commands/tmdb.rs
//! Tauri commands for TMDb metadata lookup.
//!
//! Sync `ureq` calls are dispatched on the Tokio blocking pool so the
//! UI thread keeps responding (~10s HTTP timeout). Errors come back as
//! human-readable strings the frontend can toast.

use crate::tmdb::{
    self, TmdbMovieDetails, TmdbSearchResult, TmdbTvSearchResult, TmdbTvSeasonEpisode,
};

#[tauri::command]
pub async fn tmdb_search(query: String) -> Result<Vec<TmdbSearchResult>, String> {
    tauri::async_runtime::spawn_blocking(move || tmdb::search(&query))
        .await
        .map_err(|e| format!("tmdb search join: {e}"))?
}

#[tauri::command]
pub async fn tmdb_movie_details(tmdb_id: u32) -> Result<TmdbMovieDetails, String> {
    tauri::async_runtime::spawn_blocking(move || tmdb::details(tmdb_id))
        .await
        .map_err(|e| format!("tmdb details join: {e}"))?
}

#[tauri::command]
pub async fn tmdb_tv_search(query: String) -> Result<Vec<TmdbTvSearchResult>, String> {
    tauri::async_runtime::spawn_blocking(move || tmdb::search_tv(&query))
        .await
        .map_err(|e| format!("tmdb tv search join: {e}"))?
}

#[tauri::command]
pub async fn tmdb_tv_season(
    tmdb_tv_id: u32,
    season_number: u32,
) -> Result<Vec<TmdbTvSeasonEpisode>, String> {
    tauri::async_runtime::spawn_blocking(move || {
        tmdb::tv_season(tmdb_tv_id, season_number)
    })
    .await
    .map_err(|e| format!("tmdb tv season join: {e}"))?
}
